tests/legacy/bootstrap.php

```diff
@@ -99,6 +98,58 @@ private function initialize_code_hacker() {
 		CodeHacker::enable();
 	}
 
+	/**
+	 * Re-initialize the dependency injection engine.
+	 *
+	 * The dependency injection engine has been already initialized as part of the Woo initialization, but we need
+	 * to replace the registered read-only container with a fully configurable one for testing.
+	 * To this end we hack a bit and use reflection to grab the underlying container that the read-only one stores
+	 * in a private property.
+	 *
+	 * Additionally, we replace the legacy/function proxies with mockable versions to easily replace anything
+	 * in tests as appropriate.
+	 *
+	 * @throws \Exception The Container class doesn't have a 'container' property.
+	 */
+	private function initialize_dependency_injection() {
+		try {
+			$inner_container_property = new \ReflectionProperty( \Automattic\WooCommerce\Container::class, 'container' );
+		} catch ( ReflectionException $ex ) {
+			throw new \Exception( "Error when trying to get the private 'container' property from the " . \Automattic\WooCommerce\Container::class . ' class using reflection during unit testing bootstrap, has the property been removed or renamed?' );
+		}
+
+		$inner_container_property->setAccessible( true );
+		$inner_container = $inner_container_property->getValue( wc_get_container() );
+
+		$inner_container->replace( LegacyProxy::class, MockableLegacyProxy::class );
+		$inner_container->reset_resolved();
+
+		$GLOBALS['wc_container'] = $inner_container;
+	}
+
+	/**
+	 * Register autoloader for the files in the 'tests/tools' directory, for the root namespace 'Automattic\WooCommerce\Testing\Tools'.
+	 */
+	protected static function register_autoloader_for_testing_tools() {
+		return spl_autoload_register(
+			function ( $class ) {
+				$prefix   = 'Automattic\\WooCommerce\\Testing\\Tools\\';
+				$base_dir = dirname( dirname( __FILE__ ) ) . '/Tools';
```

Comment:
You know what? This method isn't even needed. There's an entry for that in `composer.json`:

```json
  "autoload-dev": {
    "psr-4": {
      "Automattic\\WooCommerce\\Tests\\": "tests/php/src",
      "Automattic\\WooCommerce\\Testing\\Tools\\": "tests/Tools"
    }
  },
```

Reply:
I was thinking the same thing but since WooCommerce hasn't yet been initialized there isn't actually an autoloader registered to load those classes. We can't use the autoloader until after `require_once $this->wp_tests_dir . '/includes/bootstrap.php';` executes. 

With that said however, I'm not sure we want to register a custom autoloader if we don't have to. Since our only issues lie with the CodeHacker files I wonder if we can just put back the manual loading for those files and rely on WooCommerce to register its autoloader. This would make sure we're using a consistent autoloader for everything, what do you think?

Reply:
For the record, this was fixed with the two commits you added.